src/c_api/c_api_symbolic.cc

```diff
@@ -1199,3 +1200,73 @@ int MXShallowCopySymbol(SymbolHandle src, SymbolHandle* out) {
   *out = out_sym;
   API_END_HANDLE_ERROR(delete out_sym);
 }
+
+int MXOptimizeForBackend(SymbolHandle sym_handle,
+                         const char* backend_name,
+                         SymbolHandle* ret_sym_handle,
+                         const mx_uint len,
+                         NDArrayHandle* in_args_handle,
+                         const mx_uint num_options,
+                         const char** keys,
+                         const char** vals) {
+  nnvm::Symbol *s = new nnvm::Symbol();
+  API_BEGIN();
+  nnvm::Symbol *sym = static_cast<nnvm::Symbol *>(sym_handle);
+  *s = sym->Copy();
+  nnvm::Graph g = Symbol2Graph(*s);
+  if (len) {
+    NDArray **in_args_ptr = reinterpret_cast<NDArray**>(in_args_handle);
+    Context default_ctx = in_args_ptr[0]->ctx();
```

Comment:
Why we use this ctx? User may bind new ctx which is different with this, like the code you provided:
```
import mxnet as mx
sym, args, aux = mx.model.load_checkpoint('resnet-50', 0)
sym = sym.optimize_for('default', args)
exe = sym.bind(ctx=mx.cpu(), args=args, aux_states=aux, grad_req='null')
```
How to deal with default_ctx != mx.cpu()?

Reply:
InferStorageTypes is not only used by MKLDNN, but important for all backends, because it defines whether it's a sparse computation or not. Not all backends can support sparse tensor. I think we need to gather storage type along with dtype and shape for subgraph partition.

Reply:
For mkldnn backend, the correct ctx is important for getting expected result from `InferStorageTypes`, so please consider to accepts correct ctx from user. I expect the API would be like
```
sym = sym.optimize_for('default', ctx= mx.cpu(), args) #same context with bind
exe = sym.bind(ctx=mx.cpu(), args=args, aux_states=aux, grad_req='null')
```

Reply:
I don't think that check is necessary, as user can use different `args` between `optimize_for` and `bind`. 
Another case is, to support multiple backends mixed running, user may run optimize_for many times. Like,
```
sym = sym.optimize_for('MKLDNN', ctx= mx.cpu(), args) 
sym = sym.optimize_for('backendA', ctx= mx.gpu(), args) # backend A to handle the rest graph
sym = sym.optimize_for('backendB', ctx= mx.cpu(), args) # backend B to handle the rest graph
exe = sym.bind(ctx=mx.cpu(), args=args, aux_states=aux, grad_req='null')
```
For this case, context is mainly decided by graph partition, not bind.

Reply:
> Instead we'll pull the context from where the args reside.

args are normally loaded by
```
sym, args, aux = mx.model.load_checkpoint('resnet-50', 0)
```
No contexts are specified. That's why we need to pass the expected contexts during bind. The contexts from args doesn't mean anything.

>We expect that users could call this more than once for the same model. For example, users could first optimize for EIA (that would group supported nodes into subgraphs) and then optimize for MKLDNN for the remaining nodes that were not partitioned into subgraphs.

That's the comments you provided earlier. So I thought you're working to support multi-context execution. If not, then you mean we only support running multiple `optimize_for` for same contexts? Even for that, we need to pass correct contexts for `optimize_for`, not the gathered one from args.

Now again, the correct contexts are important for `InferStorageTypes` for MKLDNN backend, so we need to pass correct context(same as bind) to `optimize_for`. If you don't want to introduce `ctx` for `optimize_for` API, then I suggest to query the expected context from backend property, and use it to pass `InferStorageTypes`. Or, you can figure out other solution to solve this, I just want to ensure that subgraph partition can get same `InferStorageTypes` results as bind.

Reply:
Finally you got my point. Thanks for your patients. Either solution is OK for me, you can make the decision.